README.md
# KSUID for Ruby

This is the repository that holds all KSUID for Ruby gems. What is a [KSUID]((https://github.com/segmentio/ksuid)), you ask? The original readme for the Go library does a great job of explaining what they are and how you can use them, so we excerpt it here.

---

## What is a KSUID?

KSUID is for K-Sortable Unique IDentifier. It's a way to generate globally unique IDs similar to RFC 4122 UUIDs, but contain a time component so they can be "roughly" sorted by time of creation. The remainder of the KSUID is randomly generated bytes.

## Why use KSUIDs?

Distributed systems often require unique IDs. There are numerous solutions out there for doing this, so why KSUID?

### 1. Sortable by Timestamp

Unlike the more common choice of UUIDv4, KSUIDs contain a timestamp component that allows them to be roughly sorted by generation time. This is obviously not a strong guarantee as it depends on wall clocks, but is still incredibly useful in practice.

### 2. No Coordination Required

[Snowflake IDs][1] and derivatives require coordination, which significantly increases the complexity of implementation and creates operations overhead. While RFC 4122 UUIDv1 does have a time component, there aren't enough bytes of randomness to provide strong protections against duplicate ID generation.

KSUIDs use 128-bits of pseudorandom data, which provides a 64-times larger number space than the 122-bits in the well-accepted RFC 4122 UUIDv4 standard. The additional timestamp component drives down the extremely rare chance of duplication to the point of near physical infeasibility, even assuming extreme clock skew (> 24-hours) that would cause other severe anomalies.

[1]: https://blog.twitter.com/2010/announcing-snowflake

### 3. Lexicographically Sortable, Portable Representations

The binary and string representations are lexicographically sortable, which allows them to be dropped into systems which do not natively support KSUIDs and retain their k-sortable characteristics.

The string representation is that it is base 62-encoded, so that they can "fit" anywhere alphanumeric strings are accepted.

## How do they work?

KSUIDs are 20-bytes: a 32-bit unsigned integer UTC timestamp and a 128-bit randomly generated payload. The timestamp uses big-endian encoding, to allow lexicographic sorting. The timestamp epoch is adjusted to March 5th, 2014, providing over 100 years of useful life starting at UNIX epoch + 14e8. The payload uses a cryptographically strong pseudorandom number generator.

The string representation is fixed at 27-characters encoded using a base 62 encoding that also sorts lexicographically.

---

## Contents of this repository

Currently, there are two gems available:

1. [KSUID for Ruby](ksuid/README.md) provides the main data type and handling for KSUIDs. If you want to use them outside of a database, this will be all you need.
2. [KSUID for ActiveRecord](activerecord-ksuid/README.md) handles integrating with ActiveRecord. If you're using ActiveRecord and want to serialize KSUID columns, you will want this gem.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/) and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [0.5.0](https://github.com/michaelherold/ksuid/compare/v0.4.0...v0.5.0) - 2022-08-18

### Added

- If you'd rather deal in KSUID strings instead of `KSUID::Type`s, you can now generate them simply with `KSUID.string`. It takes the same arguments, `payload` and `time` as `KSUID.new`, but returns a string instead of a `KSUID::Type`.
- `KSUID.prefixed` and the `KSUID::Prefixed` class now can generate prefixed KSUIDs to make them visually identifiable for their source. You cannot prefix a binary-encoded KSUID, only base 62-encoded ones.
- `ActiveRecord::KSUID` now accepts a `prefix:` argument for handling prefixed KSUIDs. In addition, the `ksuid` column type also accepts a `prefix:` argument to calculate the intended size of the column with the prefix.

### Deprecated

- `KSUID::ActiveRecord` is now `ActiveRecord::KSUID`. The original constant will continue to work until v1.0.0, but will emit a warning upon boot of your application. To silence the deprecation, change all uses of `KSUID::ActiveRecord` to the new constant, `ActiveRecord::KSUID`. See the [upgrading notice][./UPGRADING.md] for more information.

### Miscellaneous

- The compatibility check for the Base62 implementation in the gem is about 10x faster now. The original optimization did not optimize as much due to an error with the benchmark. This change has a tested benchmark that shows a great improvement. Note that this is a micro-optimization and we see no real performance gain in the parsing of KSUID strings.

## [0.4.0](https://github.com/michaelherold/ksuid/compare/v0.3.0...v0.4.0) - 2022-07-29

### Added

- `KSUID::Type` acts as a proper value object now, which means that you may use it as a Hash key or use it in ActiveRecord's `.includes`. `KSUID::Type#eql?`, `KSUID::Type#hash`, and `KSUID::Type#==` now work as expected. Note that `KSUID::Type#==` is more lax than `KSUID::Type#eql?` because it can also match any object that converts to a string matching its value. This means that you can use it to match against `String` KSUIDs.

### Fixed

- `ActiveRecord::QueryMethods#include` works as expected now due to the fix on the value object semantics of `KSUID::Type`.
- Binary KSUID primary and foreign keys work as expected on JRuby.

## [0.3.0](https://github.com/michaelherold/ksuid/compare/v0.2.0...v0.3.0) - 2021-10-07

### Added

- A utility function for converting from a hexidecimal-encoded string to a byte string. This is necessary to handle the default encoding of binary fields within PostgreSQL.

## [0.2.0](https://github.com/michaelherold/ksuid/compare/v0.1.0...v0.2.0) - 2020-11-11

### Added

- The ability to configure the random generator for the gem via `KSUID.configure`. This allows you to set up random generation to the specifications you need, whether that is for speed or for security.
- Support for ActiveRecord. You can now use `KSUID::ActiveRecord[:my_field]` to define a KSUID field using the Rails 5 Attributes API. There is also two new column types for migrations: `ksuid` and `ksuid_binary`. The first stores your KSUID as a string in the database, the latter as binary data.

### Changed

- The `KSUID::Type#inspect` method now makes it much easier to see what you're looking at in the console when you're debugging.

## [0.1.0](https://github.com/michaelherold/ksuid/tree/v0.1.0) - 2017-11-05

### Added

- Basic `KSUID.new` interface.
- Parsing of bytes through `KSUID.from_bytes`.
- Parsing of strings through `KSUID.from_base62`.


ksuid/lib/ksuid.rb
# frozen_string_literal: true

require_relative 'ksuid/configuration'
require_relative 'ksuid/version'

# The K-Sortable Unique IDentifier (KSUID)
#
# Distributed systems require unique identifiers to track events throughout
# their subsystems. Many algorithms for generating unique identifiers, like the
# {https://blog.twitter.com/2010/announcing-snowflake Snowflake ID} system,
# require coordination with a central authority. This is an unacceptable
# constraint in the face of systems that run on client devices, yet we still
# need to be able to generate event identifiers and roughly sort them for
# processing.
#
# The KSUID optimizes this problem into a roughly sortable identifier with
# a high possibility space to reduce the chance of collision. KSUID uses
# a 32-bit timestamp with second-level precision combined with 128 bytes of
# random data for the "payload". The timestamp is based on the Unix epoch, but
# with its base shifted forward from 1970-01-01 00:00:00 UTC to 2014-05-13
# 16:532:20 UTC. This is to extend the useful life of the ID format to over
# 100 years.
#
# Because KSUID timestamps use seconds as their unit of precision, they are
# unsuitable to tasks that require extreme levels of precision. If you need
# microsecond-level precision, a format like {https://github.com/alizain/ulid
# ULID} may be more suitable for your use case.
#
# KSUIDs are "roughly sorted". Practically, this means that for any given event
# stream, there may be some events that are ordered in a slightly different way
# than they actually happened. There are two reasons for this. Firstly, the
# format is precise to the second. This means that two events that are
# generated in the same second will be sorted together, but the KSUID with the
# smaller payload value will be sorted first. Secondly, the format is generated
# on the client device using its clock, so KSUID is susceptible to clock shift
# as well. The result of sorting the identifiers is that they will be sorted
# into groups of identifiers that happened in the same second according to
# their generating device.
#
# @example Generate a new KSUID
#   KSUID.new
#
# @example Generate a KSUID prefixed by `evt_`
#   KSUID.prefixed('evt_')
#
# @example Parse a KSUID string that you have received
#   KSUID.from_base62('aWgEPTl1tmebfsQzFP4bxwgy80V')
#
# @example Parse a KSUID byte string that you have received
#   KSUID.from_bytes(
#     "\xFF\xFF\xFF\xFF\xFF\xFF\xFF\xFF\xFF\xFF\xFF\xFF\xFF\xFF\xFF\xFF\xFF\xFF\xFF\xFF"
#   )
#
# @example Parse a KSUID byte array that you have received
#   KSUID.from_bytes(
#     [255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255,
#      255, 255, 255, 255]
#   )
module KSUID
  # The shift in the Unix epoch time between the standard and the KSUID base
  #
  # @return [Integer] the number of seconds by which we shift the epoch
  EPOCH_TIME = 1_400_000_000

  # The number of bytes that are used to represent each part of a KSUID
  #
  # @return [Hash{Symbol => Integer}] the map of data type to number of bytes
  BYTES = { base62: 27, payload: 16, timestamp: 4, total: 20 }.freeze

  # The number of characters in a base 62-encoded KSUID
  #
  # @return [Integer]
  STRING_LENGTH = 27

  # The maximum KSUID as a base 62-encoded string.
  #
  # @return [String]
  MAX_STRING_ENCODED = 'aWgEPTl1tmebfsQzFP4bxwgy80V'

  autoload :Base62, 'ksuid/base62'
  autoload :Prefixed, 'ksuid/prefixed'
  autoload :Type, 'ksuid/type'
  autoload :Utils, 'ksuid/utils'

  # Converts a KSUID-compatible value into an actual KSUID
  #
  # @api public
  #
  # @example Converts a base 62 KSUID string into a KSUID
  #   KSUID.call('15Ew2nYeRDscBipuJicYjl970D1')
  #
  # @param ksuid [String, Array<Integer>, KSUID::Type] the KSUID-compatible value
  # @return [KSUID::Type] the converted KSUID
  # @raise [ArgumentError] if the value is not KSUID-compatible
  def self.call(ksuid)
    return unless ksuid

    case ksuid
    when KSUID::Prefixed then ksuid.to_ksuid
    when KSUID::Type then ksuid
    when Array then KSUID.from_bytes(ksuid)
    when String then cast_string(ksuid)
    else
      raise ArgumentError, "Cannot convert #{ksuid.inspect} to KSUID"
    end
  end

  # The configuration for creating new KSUIDs
  #
  # @api private
  #
  # @return [KSUID::Configuration] the gem's configuration
  def self.config
    @config ||= KSUID::Configuration.new
  end

  # Configures the KSUID gem by passing a block
  #
  # @api public
  #
  # @example Override the random generator with a null data generator
  #   KSUID.configure do |config|
  #     config.random_generator = -> { "\x00" * KSUID::BYTES[:payload] }
  #   end
  #
  # @example Override the random generator with the faster, but less secure, Random
  #   KSUID.configure do |config|
  #     config.random_generator = -> { Random.new.bytes(KSUID::BYTES[:payload]) }
  #   end
  #
  # @return [KSUID::Configuration] the gem's configuration
  def self.configure
    yield config if block_given?
    config
  end

  # Converts a base 62-encoded string into a KSUID
  #
  # @api public
  #
  # @example Parse a KSUID string into an object
  #   KSUID.from_base62('0vdbMgWkU6slGpLVCqEFwkkZvuW')
  #
  # @param string [String] the base 62-encoded KSUID to convert into an object
  # @return [KSUID::Type] the KSUID generated from the string
  def self.from_base62(string)
    string = string.rjust(STRING_LENGTH, Base62::CHARSET[0]) if string.length < STRING_LENGTH
    int = Base62.decode(string)
    bytes = Utils.int_to_bytes(int, 160)

    from_bytes(bytes)
  end

  # Converts a byte string or byte array into a KSUID
  #
  # @api public
  #
  # @example Parse a KSUID byte string into an object
  #   KSUID.from_bytes("\x06\x83\xF7\x89\x04\x9C\xC2\x15\xC0\x99\xD4+xM\xBE\x994\e\xD7\x9C")
  #
  # @param bytes [String, Array<Integer>] the byte string or array to convert into an object
  # @return [KSUID::Type] the KSUID generated from the bytes
  def self.from_bytes(bytes)
    bytes = bytes.bytes if bytes.is_a?(String)

    timestamp = Utils.int_from_bytes(bytes.first(BYTES[:timestamp]))
    payload = Utils.byte_string_from_array(bytes.last(BYTES[:payload]))

    KSUID::Type.new(payload: payload, time: Time.at(timestamp + EPOCH_TIME))
  end

  # Generates the maximum KSUID as a KSUID type
  #
  # @api semipublic
  #
  # @example Generate the maximum KSUID
  #   KSUID.max
  #
  # @return [KSUID::Type] the maximum KSUID in both timestamp and payload
  def self.max
    from_bytes([255] * BYTES[:total])
  end

  # Instantiates a new KSUID
  #
  # @api public
  #
  # @example Generate a new KSUID for the current second
  #   KSUID.new
  #
  # @example Generate a new KSUID for a given timestamp
  #   KSUID.new(time: Time.parse('2017-11-05 15:00:04 UTC'))
  #
  # @param payload [String, Array<Integer>, nil] the payload for the KSUID
  # @param time [Time] the timestamp to use for the KSUID
  # @return [KSUID::Type] the generated KSUID
  def self.new(payload: nil, time: Time.now)
    Type.new(payload: payload, time: time)
  end

  # Instantiates a new {KSUID::Prefixed}
  #
  # @api public
  # @since 0.5.0
  #
  # @example Generate a new prefixed KSUID for the current second
  #   KSUID.prefixed('evt_')
  #
  # @example Generate a new prefixed KSUID for a given timestamp
  #   KSUID.prefixed('cus_', time: Time.parse('2022-08-16 10:36:00 UTC'))
  #
  # @param prefix [String] the prefix to apply to the KSUID
  # @param payload [String, Array<Integer>, nil] the payload for the KSUID
  # @param time [Time] the timestamp to use for the KSUID
  # @return [KSUID::Prefixed] the generated, prefixed KSUID
  def self.prefixed(prefix, payload: nil, time: Time.now)
    Prefixed.new(prefix, payload: payload, time: time)
  end

  # Generates a KSUID string
  #
  # @api public
  # @since 0.5.0
  #
  # @example Generate a new KSUID string for the current second
  #   KSUID.string
  #
  # @example Generate a new KSUID string for a given timestamp
  #   KSUID.string(time: Time.parse('2017-11-05 15:00:04 UTC'))
  #
  # @param payload [String, Array<Integer>, nil] the payload for the KSUID string
  # @param time [Time] the timestamp to use for the KSUID string
  # @return [String] the generated string
  def self.string(payload: nil, time: Time.now)
    Type.new(payload: payload, time: time).to_s
  end

  # Casts a string into a KSUID
  #
  # @api private
  #
  # @param ksuid [String] the string to convert into a KSUID
  # @return [KSUID::Type] the converted KSUID
  def self.cast_string(ksuid)
    if Base62.compatible?(ksuid)
      KSUID.from_base62(ksuid)
    else
      KSUID.from_bytes(ksuid)
    end
  end
  private_class_method :cast_string
end
